File: back/node/app/delivery/feature/game/Launcher.ts
import { createRedisClient } from "../../../connections/infrastructure/database/redis.js";
import { RedisData } from "../../../core/database/redis/RedisData.js";
import {GameRepositoryImpl} from "../../../core/data/implementation/redis/GameRepositoryImpl.js";
import {GameStateChecker} from "../../../core/service/game/GameStateChecker.js";
import {GameLogic} from "../../../core/service/game/GameLogic.js";
import {GameUseCases} from "../../../core/domain/GameUseCases.js";
import {GameResponses} from "../../utils/responses/SResponses.js";
import {GameController} from "./controllers/GameController.js";
import {GameEndpoints} from "./endpoints/GameEndpoints.js";
import { createExpressRouter, createExpressServer } from "../../../connections/infrastructure/server/express.js";

const lauchGameFeature = async () =>{

    // Redis Client & Connect =================================================================
    const redisClient = createRedisClient();
    await redisClient.connect();

    const redisData = new RedisData(redisClient)
    const gameRepository = new GameRepositoryImpl(redisData)

    const gameChecker = new GameStateChecker()
    const gameLogic = new GameLogic(gameChecker)

    const gameUseCases = new GameUseCases(gameRepository, gameLogic)
    const gameResponses = new GameResponses()

    const gameController = new GameController(gameUseCases, gameResponses)

    // Express Router + HTTP Routes
    const router = createExpressRouter();

    new GameEndpoints(gameController, router);

    // Start Express Server ============================================================================
    createExpressServer([], router);

    // Ensure you call `client.quit()` when you are done with Redis
    process.on('exit', () => {
        redisClient.quit().then(() => {
            console.log('Redis client quit successfully');
            process.exit(0);
        }).catch((err: unknown) => {
            console.error('Error quitting Redis client:', err);
            process.exit(1);
        });
    });

    process.on('uncaughtException', (err) => {
        console.error('There was an uncaught error', err)
        process.exit(1) //mandatory (as per the Node.js docs)
    })

}

export {lauchGameFeature}

File: back/kotlin/app/delivery/feature/game/Launcher.kt
package delivery.feature.game

import io.lettuce.core.RedisClient
import io.lettuce.core.api.sync.RedisCommands
import io.lettuce.core.api.async.RedisAsyncCommands
import io.ktor.server.application.*
import io.ktor.server.routing.*

import core.database.redis.RedisData
import core.data.implementation.redis.GameRepositoryImpl
import core.service.game.GameStateChecker
import core.service.game.GameLogic
import core.domain.GameUseCases
import delivery.utils.responses.GameResponses
import delivery.feature.game.controllers.GameController
import delivery.feature.game.endpoints.GameEndpoints


fun launchGameFeature(application: Application, redisData: RedisData){

    val gameRepository = GameRepositoryImpl(redisData)

    val gameChecker = GameStateChecker()
    val gameLogic = GameLogic(gameChecker)

    val gameUseCases = GameUseCases(gameRepository, gameLogic)
    val gameResponses = GameResponses()

    val gameController = GameController(gameUseCases, gameResponses)

    val gameRoutes = GameEndpoints(gameController, application).gameRoutes()

}